File: AdventOfCode/2022/8/day8part1.cpp
#include <iostream>
#include <fstream>
#include <string>
using namespace std;

int main() {
    string filename = "example";
    ifstream in (filename);
    if (in.is_open()) {
        int size = 0;
        string tmp;
        getline(in, tmp);
        size = tmp.size(); // hopefully its square

        string arr[size];
        arr[0] = tmp;
        for (int i = 1; i < size; i++) {
            getline(in, arr[i]);
            // cout << arr[i] << endl;
        }

        int count = 0;
        for (int i = 1; i < size-1; i++) {
            for (int j = 1; j < size-1; j++) {
                int current = arr[i][j] - '0';

                bool visible = true;
                for (int x = 0; x < j; x++) { // left
                    if (current <= arr[i][x] - '0') {
                        cout << x << i << ' ';
                        cout << current << ' ' << arr[i][x] - '0';
                        visible = false;
                        break;
                    }
                }
                cout << endl;
                if (!visible) { // don't need to check if it was not visible
                    for (int x = j + 1; x < size; x++) { // right
                        if (current <= arr[i][x] - '0') {
                            visible = false;
                            break;
                        }
                    }
                }
                if (!visible) {
                    for (int y = 0; y < i; y++) { // top
                        if (current <= arr[y][j] - '0') {
                            visible = false;
                            break;
                        }
                    }
                }
                if (!visible) {
                    for (int y = i + 1; y < size; y++) { // bottom
                        if (current <= arr[y][j] - '0') {
                            visible = false;
                            break;
                        }
                    }
                }

                if (visible) {
                    count++;
                }
            }
        }
        cout << count << endl;
    }
}

File: AdventOfCode/2022/8/day8part1.py
eg = "example"
inp = "input"
test = "test"
filename = inp
f = open(filename, "r")
contents = f.read().split("\n")
size = len(contents[0])

def to_arr(old):
	new_arr = []
	for r in old:
		new_sub_arr = []
		for c in r:
			new_sub_arr.append(int(c))
		new_arr.append(new_sub_arr)
	return new_arr

arr = to_arr(contents)

visible = (size - 1) * 4

for i in range(1, size-1): # rows (553)
	for j in range(1, size-1): # columns (551)
		tree = arr[i][j]
		is_visible = True

		"""
		1,1 visible top, left
		1,2 visible top, right
		2,1 visible right
		2,3 visible right
		3,2 visible bottom, left
		"""

		# top
		for x in range(0, i):
			if arr[x][j] >= tree:
				is_visible = False

		# bottom
		if not is_visible:
			is_visible = True
			for x in range(i+1, size):
				if arr[x][j] >= tree:
					is_visible = False

		# left
		if not is_visible:
			is_visible = True
			for x in range(0, j):
				if arr[i][x] >= tree:
					is_visible = False

		# right
		if not is_visible:
			is_visible = True
			for x in range(j+1, size):
				if arr[i][x] >= tree:
					is_visible = False

		if is_visible:
			print("visible")
			visible += 1

		# if not (i == 1 and j == 2\
		# 	or i == 2 and j == 1\
		# 	or i == 2 and j == 3):
		# 	continue

print(visible)